### PR TITLE
Improvements to classification perf with large string literals.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Followup to https://github.com/dotnet/roslyn/pull/72216.

Takes classification cost down from:

image

to

image

Basically, we were producign *lots* of tags in the case of some very large strings.  This ended up being costly with later processing.  This change fixes us up to produce far fewer tags (gated to just what is in view).  

--

There is still further optimizations we can do.  Specifically, for regex/json classification, we can limit how we walk those language subtrees to avoid even classifying portions of the embedded language tree that is not in view.

### CODE REVIEW

File: src/Features/Core/Portable/EmbeddedLanguages/Classification/AbstractEmbeddedLanguageClassificationService.cs [hunk 1]
```diff
@@ -106,17 +106,20 @@ public void VisitTokens(SyntaxNode node)
...
             {
                 _cancellationToken.ThrowIfCancellationRequested();
-                ProcessTriviaList(token.LeadingTrivia);
+
+                // Directives need to be processes as they can contain strings, which then have escapes in them.
+                if (token.ContainsDirectives)
+                    ProcessTriviaList(token.LeadingTrivia);
+
                 ClassifyToken(token);
-                ProcessTriviaList(token.TrailingTrivia);
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EmbeddedLanguages/Classification/AbstractEmbeddedLanguageClassificationService.cs:111]: we were also previous walking down into trailing trivia (which never has structure).  we were also walking into *all* structure, even though we don't have any embedded-lang classification in structured-trivia *except* for the case of `string literals` in directives.  so this means no more walking into things like doc comments, which is nice.

File: src/EditorFeatures/Core/Classification/Semantic/ClassificationUtilities.cs [hunk 2]
```diff
@@ -123,16 +123,28 @@ public static TagSpan<IClassificationTag> Convert(IClassificationTypeMap typeMap
                 return false;
             }
 
-            var subTextSpan = service.GetMemberBodySpanForSpeculativeBinding(member);
-            if (subTextSpan.IsEmpty)
+            var memberBodySpan = service.GetMemberBodySpanForSpeculativeBinding(member);
+            if (memberBodySpan.IsEmpty)
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Core/Classification/Semantic/ClassificationUtilities.cs:127]: view with whitespace off.

File: src/Features/Core/Portable/EmbeddedLanguages/Classification/EmbeddedLanguageClassifierContext.cs [hunk 3]
```diff
@@ -8,52 +8,63 @@
...
-            VirtualCharService = virtualCharService;
-            _result = result;
-            CancellationToken = cancellationToken;
-        }
-
-        public void AddClassification(string classificationType, TextSpan span)
-            => _result.Add(new ClassifiedSpan(classificationType, span));
+        // Ignore characters that don't intersect with the requested span.  That avoids potentially adding lots of
+        // classifications for portions of a large string that are out of view.
+        if (span.IntersectsWith(_spanToClassify))
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/EmbeddedLanguages/Classification/EmbeddedLanguageClassifierContext.cs:67]: note: in the case the user provided, while the string contains json, it's not a JsonTree.  All we ahve are the embedded classifications for escapes like `""` in the middle of the string.  But there are literally thousands of these classifications.  So this takes us down from many thousands of classifications to a few dozen.

**Comment by CyrusNajmabadi** [comment]
@sharwell @ToddGrun this is ready for review.

**Review by ToddGrun — APPROVED** [review]
LGTM

**Pull request merged**